### DISCUSSION TITLE
Server function not working properly

### DISCUSSION THREAD

**Discussion by Neighborkid01** [leptos-rs/leptos#1229] [Q&A] [answered]
Hello, I'm pretty new to Leptos and Rust in general and I've run into a bit of a head-scratcher and Google isn't turning anything up for me...
I have an app that's using `leptos_axum` for my rendering, I'm using the `ssr` feature, and I've got `sqlx` hooked up to a Postgres database (not sure what part of that info is relevant).

I had a basic skeleton working where I just had hard-coded `Vec`s of ids and I could navigate between the `/feeds` page and a specific `/feeds/:id` page just fine.
On the `/feeds` page, I have a link that takes you to each `/feed/:id` page, and on those pages I have a link back to `/feeds`. I  also have another link to `/feeds` in my nav bar.
I just got my db connected and added a server function defined like this
```RUST
#[server(GetFeeds, "/api")]
pub async fn get_feeds() -> Result<Vec<Feed>, ServerFnError>
```

Now that I have this function to query my database, I'm using `create_resource()` for the records I pull out of the database.
After this change, if I click the link to `/feeds` in my nav bar, it works as expected, but once I've navigated to one of my feed detail pages, instead of the links to `/feeds` making a `GET`, they're making a `POST` to `/api/get_feeds8468097346202768843` (or some other random hash).
I've tried changing my `get_feeds` function to have `#[server(GetFeeds, "/api", "GetJson")]`, but I got this compile error:
```
error: custom attribute panicked
  --> src/pages/feeds_page.rs:52:1
   |
52 | #[server(GetFeeds, "/api", "GetJson")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: message: proc-macro-error API cannot be used outside of `entry_point` invocation, perhaps you forgot to annotate your #[proc_macro] function with `#[proc_macro_error]
```

Any help would be greatly appreciated, thanks!

**Comment by gbj** [comment]
Hi! This is not quite enough information to understand what's going on.

A few notes:
1. These are not routes, they are server function URLs, which have a hash added for disambiguation from one another. (It's based on their location in your app.)
2. Server functions default to POST, so what you're describing is completely expected.
3. I'm not sure what the issue is with the `"GetJson"` server function macro invocation, there, but I'm not able to reproduce it, either. 

What's the nature of the question? Is there something that's not working in the app? Are you wondering why the POST request and hashes are used?

A more full example might be helpful, especially for the server macro issue.

**Reply by Neighborkid01** [reply]
Thanks for the clarification. I've updated the title to better reflect my issue.

My main problem is that for some reason, the first time my `/feeds` page loads, it renders correctly and the 2 entries in my db are shown, and I can see in my server logs where the records are getting queried, but whenever I redirect there from one of my links, I get this error in the browser console:
```
Could not find a server function at the route get_feeds8468097346202768843. 

It's likely that you need to call ServerFn::register() on the server function type, somewhere in your `main` function.
```

I have this in my main function, which I believe should be registering my server function (seeing as how it works the first time)
```RUST
let conf = get_configuration(None).await.unwrap();
let leptos_options = conf.leptos_options;
let addr = leptos_options.site_addr;
let routes = generate_route_list(|cx| view! { cx, <App/> }).await;

let app = Router::new()
  .route("/api/*fn_name", post(leptos_axum::handle_server_fns))
  .leptos_routes(leptos_options.clone(), routes, |cx| view! { cx, <App/> })
  .fallback(file_and_error_handler)
  .layer(Extension(Arc::new(leptos_options)));
```

Also, here's my complete `feeds_page.rs` and `feed_page.rs` since I'm not sure where the issue lies.
I'm not sure fi maybe I'm not importing something properly and that's what's causing the macro to not work.

`feeds_page.rs`
```RUST
use leptos::*;
use leptos_router::*;
use crate::models::feed::Feed;
use crate::error_template::{ErrorTemplate, ErrorTemplateProps};

#[cfg(feature = "ssr")]
use crate::db::*;

/// Renders the Feeds page for a user
#[component]
pub fn FeedsPage(cx: Scope) -> impl IntoView {
  // list of feeds is loaded from the server in reaction to changes
  let feeds = create_resource(
    cx,
    || (),
    |_| async move { get_feeds().await },
  );

  view! { cx,
    <h1>"Feeds"</h1>
    <ul>
      <Transition fallback=move || view! {cx, <p>"Loading..."</p> }>
        <ErrorBoundary fallback=|cx, errors| view!{cx, <ErrorTemplate errors=errors/>}>
          {feeds.read(cx).map(move |feeds| {
            match feeds {
              Err(e) => { view! { cx, <pre class="error">"Server Error: " {e.to_string()}</pre>}.into_view(cx) },
              Ok(feeds) => {
                if feeds.is_empty() { return view! { cx, <p>"No feeds were found."</p> }.into_view(cx); }

                feeds.into_iter().map(move |feed|
                  view! { cx, <FeedLink feed=feed /> }
                ).collect::<Vec<_>>().into_view(cx)
              }
            }
          }).unwrap_or_default()}
        </ErrorBoundary>
      </Transition>
    </ul>
  }
}

/// Renders a single feed link
#[component]
pub fn FeedLink(cx: Scope, feed: Feed) -> impl IntoView {
  view! { cx,
    <li>
      <A href=format!("{}", feed.feed_code)>{format!("Feed {}", feed.name)}</A>
    </li>
  }
}

/// Retrieves all feeds from the database
#[server(GetFeeds, "/api")]
pub async fn get_feeds() -> Result<Vec<Feed>, ServerFnError> {
  let mut conn = db().await?;

  let feeds = sqlx::query_as!(Feed, "SELECT * FROM feeds")
    .fetch_all(&mut conn)
    .await
    .map_err(|e| ServerFnError::ServerError(e.to_string()))?;

  Ok(feeds)
}
```

`feed_page.rs`
```RUST
use leptos::*;
use leptos_router::*;

/// Renders the page for a specific feed
#[component]
pub fn FeedPage(cx: Scope) -> impl IntoView {
  let params = use_params_map(cx);
  let feed_code = move || { params.with(|params| params.get("feed_code").cloned()) };
  view! { cx,
    <h1>"Feed " {feed_code}</h1>

    <A href="/feeds">"Back to Feeds"</A>
  }
}
```

Here is a video of my issue happening with the console error I'm getting as well:

https://github.com/leptos-rs/leptos/assets/60106973/f3f64e2b-5f0e-4cd5-8d37-2ad357105911

**Reply by Neighborkid01** [reply]
Welp...

Adding this to main fixed it for me.
```RUST
let _ = pages::feeds_page::GetFeeds::register();
```

Is there a way to register my server functions programmatically, or do I need to do each one I write like this?
Thanks!

**Reply by gbj** [reply]
They need to be registered manually in 0.3.x, and are registered automatically in the git main branch and future released versions.

**Reply by gbj** [reply, marked as answer]
Just adding for the sake of knowledge and for any future readers: the reason it worked on initial load but not subsequently was that, during server-side rendering, server functions are actually just called directly on the server, as functions. The "registration" process is required to provide each one with a distinct URL that the server knows about, but these separate server function URLs are only used when a server function is called from the browser, not when it's called on the server (i.e., during server rendering.) So a server function worked properly on initial SSR of the page, but not if you navigated to it again from the browser.

As I mentioned, this registration is now done automatically (#1154), a new version just hasn't been released yet because of some other changes on the roadmap.

If you're looking at the examples, note that they always match the git main version, which means they may have features (like automatic registration) not yet present on 0.3. [See here](https://github.com/leptos-rs/leptos/tree/v0.3.0/examples) for examples as of the `0.3.0` release.
